lib/shepherd.js

```diff
@@ -428,97 +418,109 @@ ZShepherd.prototype.forceRemove = function(ieeeAddr, callback) {
 	});
 };
 
-ZShepherd.prototype.generateMap = function(opts, callback){
-	var info = this.info();
-	var self = this;
-	var counter = 1;
-	var learn;
+ZShepherd.prototype.recursiveDeviceLQI = function(ieeeAddr, nodes, links, callback) {
+	var _nodes = nodes || {};
+	var _links = links || [];
+	var _queue = [];
 
-	var queue = []
-	var done = [];
+	debug.shepherd('perform LQI for', ieeeAddr);
 
-	var links = [];
+	// Get all devices connected to this device
+	this.quickLQI(ieeeAddr)
+		.then(function(result) {
+			if (Array.isArray(result)) {
+				_nodes[ieeeAddr] = result;
 
-	var emitRecursive = function(device,ieeeAddr,nwkAddr){
-		done.push(ieeeAddr);
-		if(ieeeAddr != info.net.ieeeAddr){
-			//self.emit('ind', { type: 'lqiFound', data: ieeeAddr, endpoints: endpoints});
+				debug.shepherd('LQI', ieeeAddr, 'result', result);
+
+				// Loop all devices and if one is discovered that has not been LQI'd before do it
+				result.forEach(function (node) {
+					if (typeof _nodes[node.ieeeAddr] === 'undefined') {
+						_queue.push(function (callback) {
+							this.recursiveDeviceLQI(node.ieeeAddr, _nodes, _links, callback);
+						}.bind(this));
+					}
+				}.bind(this));
 
-			var dev = self._findDevByAddr(ieeeAddr);
-			if(dev){
-				if (dev.status === 'offline'){
-					console.log('$$$$ LQI DEV: '+ieeeAddr);
-					self.controller.emit('ZDO:endDeviceAnnceInd', { srcaddr: nwkAddr, nwkaddr: nwkAddr, ieeeaddr: ieeeAddr, capabilities: {} });
+				if (_queue.length === 0) {
+					debug.shepherd('LQI', ieeeAddr, 'done');
+					return callback(null, true)
 				}
+				debug.shepherd('execute queue', _queue.length);
+				// Perform all request sequentially
+				return async.series(_queue, function () {
```

Comment:
Decided to leave it in, Q doesn't have an `async.series` alternative, and the recursive algorithm depends on passing on the callback, meaning it would require me to rewrite it again.